url7d636e8121_laforge49_JActor2_tgz-pJ8-org_agilewiki_jactor2_core_revisited_EHJ8/src/org/agilewiki/jactor2/core/impl/mtPlant/ReactorPoolThreadManager.java

```diff
@@ -0,0 +1,173 @@
+        this.threadCount = _threadCount;
+        final Runnable runnable = new Runnable() {
+            @Override
+            public void run() {
+                final ReactorPoolThread currentThread = (ReactorPoolThread) Thread
+                        .currentThread();
+                while (true) {
+                    try {
+                        taskRequest.acquire();
+                        PoolThreadReactorMtImpl reactor = reactors.poll();
+                        if (reactor != null) {
+                            AtomicReference<Thread> threadReference = reactor
+                                    .getThreadReference();
+                            if ((threadReference.get() == null)
+                                    && threadReference.compareAndSet(null,
+                                    currentThread)) {
+                                currentThread.setCurrentReactor(reactor);
+                                currentThread.clearMigrationCount();
+                                while (true) {
+                                    try {
+                                        reactor.run();
+                                    } catch (final MigrationException me) {
+                                        final boolean hasWork = reactor
+                                                .hasWork();
+                                        threadReference.set(null);
+                                        if (reactor.isIdler() || hasWork
+                                                || reactor.hasConcurrent()) {
+                                            execute(reactor);
+                                        }
+                                        reactor = me.reactor;
+                                        threadReference = reactor
+                                                .getThreadReference();
+                                        currentThread
+                                                .setCurrentReactor(reactor);
+                                        continue;
+                                    } catch (final Throwable e) {
+                                        logger.error(
+                                                "Exception thrown by a targetReactor's run method",
+                                                e);
+                                    }
+                                    final boolean hasWork = reactor.hasWork();
+                                    threadReference.set(null);
+                                    if (hasWork || reactor.hasConcurrent()) {
+                                        execute(reactor);
+                                    }
+                                    break;
+                                }
+                                currentThread.setCurrentReactor(null);
+                            }
+                        }
+                    } catch (final InterruptedException e) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: InterruptedException is ignored. This can delay thread shutdown and clear the thread’s interrupt status. Only code that implements a thread’s interruption policy can swallow an interruption request.

**Fix**: Rethrow the InterruptedException or reinterrupt the current thread using *Thread.currentThread().interrupt()* so that higher-level interrupt handlers can function correctly.
 If you are wrapping the InterruptedException inside a RuntimeException, call *Thread.currentThread().interrupt()* before throwing the RuntimeException.

Learn more about [interrupts](https://docs.oracle.com/javase/tutorial/essential/concurrency/interrupt.html) and [dealing with InterruptedException](https://www.ibm.com/developerworks/java/library/j-jtp05236/index.html?ca=drs-#N10187)